src/verilog/verilog_typecheck_expr.cpp

```diff
@@ -2146,6 +2146,32 @@ typet verilog_typecheck_exprt::enum_decay(const typet &type)
 
 /*******************************************************************\
 
+Function: verilog_typecheck_exprt::union_decay
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void verilog_typecheck_exprt::union_decay(exprt &expr) const
+{
+  // 1800-2017 7.3.1
+  // Verilog union types decay to a vector type [$bits(t)-1:0] when
+  // used in relational or arithmetic expressions.
+  auto &type = expr.type();
+  if(type.id() == ID_union)
+  {
+    auto new_type =
+      unsignedbv_typet{numeric_cast_v<std::size_t>(get_width(type))};
+    expr = typecast_exprt{expr, new_type};
```

Comment:
Will this reach our back-end? I don't think we'd support casts from a union. I suspect we'd need an extractbits here instead.

Reply:
Yes, the synthesis stage turns these into a member expression -- will add comment.